Pagination

brandur opened the pull request:
Usage on a top-level collection:

```
Stripe::Customer.list.each_with_all_pages do |customer|
  puts customer
end
```

Usage on a subcollection:

``` ruby
customer.invoices.each_with_all_pages do |invoice|
  puts invoice
end
```

We've also renamed `#all` to `#list` to prevent confusion ("all" implies
that all resources are being returned, and in Stripe's paginated API
this was not the case). An alias has been provided for backward API
compatibility.

Still considering the naming of `#each_with_all_pages`. It's a little
unwieldy.

Fixes #167.

Replaces #211 and #248.

/cc @kyleconroy @russelldavis Would love your input on this! Thanks!

brandur commented:
Ugh. While moving tests off of the deprecated `#all` method, I noticed that file uploads had implemented by just copy and pasting the method body from `Stripe::APIOperations::List`. In an effort to get this back on mainline, I added a class-level `api_url` configuration on API resources (similar to the existing "url", which should probably actually be "path") which file uploads can use to configure this parameter and still use the implementation from `Stripe::APIOperations::List`.

Unfortunately, the side effect is an overly large diff. I kind of want to take another pass after this is merged and try to clean up a few more things; the number of interdependent modules in this project is getting ridiculous.

russelldavis commented on lib/stripe/list_object.rb:
```diff
@@ -12,10 +22,35 @@ def [](k)
+    # Note that this method will make as many API calls as necessary to fetch
+    # all resources. For more granular control, please see +each+ and
+    # +next_page+.
+    def each_with_all_pages(&blk)
```
As currently written, if you pass a block to this, it won't actually call it. I think you want something like this:

```
def each_with_all_pages
  return enum_for(:each_with_all_pages) unless block_given?
  page = self
  loop do
    page.each do |resource|
      yield resource
    end

    page = page.next_page
    break if page.empty?
  end
end
```

(And should also add some tests that call this with a block.)

brandur replied:
@russelldavis Dang it, you're totally right. Fixed in bfc1e5b.

russelldavis commented:
Not sure if this was an intentional change or not, but I think the name makes more sense as `each_in_all_pages` (or maybe `each_from_all_pages`) rather than `each_with_all_pages` (`with` typically implies another argument being passed to the block, as in [each_with_object](http://ruby-doc.org/core-2.2.3/Enumerable.html#method-i-each_with_object)).

Another option here would be `auto_paging_each`.

russelldavis commented:
I'm actually starting to lean towards `auto_paging_each`, as it might be a little more clear that paging will be happening *during* iteration (i.e., there might be a long delay between yielding item 100 and 101). This name adapts pretty well to other languages' naming schemes as well (e.g., `auto_paging_iter`)

brandur commented:
@russelldavis Good call! I actually don't feel too strongly about this one, but I like your second justification. Renamed to `auth_paging_each` in c488950.

Thanks for the review!

brandur commented:
And BTW, current deployment plan is to cut a version of the gem before pulling this in ... I'd like to get a little refactoring in before putting this in its own version.

russelldavis commented on lib/stripe/list_object.rb:
```diff
@@ -12,10 +30,35 @@ def [](k)
+    # Note that this method will make as many API calls as necessary to fetch
+    # all resources. For more granular control, please see +each+ and
+    # +next_page+.
+    def auto_paging_each(&blk)
```
Nit: no need for `&blk` here since we're just using `yield`. Alternatively, you could keep `&blk` here and change lines 53-55 to `page.each(&blk)`

brandur replied:
@russelldavis I personally really like having the `(&blk)` on the end of the method because it makes it emminently obvious when you're reading through code that this is a method that takes a block. I've changed the implementation to respect that variable in 53b214f as you suggested. Looks very, very clean IMO. Excellent call!

iloveitaly commented:
:+1: Really excited about this change and the other improvements to the gem. Awesome work, @brandur!

brandur commented:
@iloveitaly Yes, I also can't wait to see this gem in a good state! Thanks for the kind words :)

brandur commented:
Removed the addition of `.api_url` and `#api_url` to `APIResource` in 58fe5ae because Kyle's approach in #316 is quite a bit more efficient and gets the same result.

Rebased against master and squashed.
